File: AI/middleware/memory_summarization.py
"""
第三层: 记忆压缩 (MemorySummarizationMiddleware)

问题场景:
- 智能体记忆无限增长,导致上下文窗口溢出
- 旧记忆占用过多空间

解决方案:
- 定期压缩旧记忆
- 保留重要记忆
"""

from typing import Any, List
import random

from .base import BaseMiddleware, MiddlewareContext, MiddlewareResult


class MemorySummarizationMiddleware(BaseMiddleware):
    """记忆压缩中间件"""

    def __init__(self,
                 max_memories: int = 100,
                 keep_recent: int = 20,
                 importance_threshold: float = 0.7):
        super().__init__("MemorySummarizationMiddleware")
        self.max_memories = max_memories
        self.keep_recent = keep_recent
        self.importance_threshold = importance_threshold

    def process(self, context: MiddlewareContext, agent: Any) -> MiddlewareResult:
        if not self.enabled:
            return MiddlewareResult()

        # Check if agent has memory
        if hasattr(agent, 'memory'):
            memories = agent.memory
            memory_count = len(memories) if hasattr(memories, '__len__') else 0

            if memory_count > self.max_memories:
                self._compress_memories(agent)

        return MiddlewareResult()

    def _compress_memories(self, agent: Any):
        """压缩记忆"""
        memories = agent.memory
        if not hasattr(memories, 'memories'):
            return

        mem_list = memories.memories
        if len(mem_list) <= self.max_memories:
            return

        # Keep recent + important memories
        recent = mem_list[-self.keep_recent:]
        old = mem_list[:-self.keep_recent]

        important = [m for m in old if getattr(m, 'importance', 0.5) >= self.importance_threshold]

        # Random sample if still too many
        remaining = max(0, self.max_memories - len(recent) - len(important))
        if len(old) - len(important) > remaining:
            others = random.sample(
                [m for m in old if getattr(m, 'importance', 0.5) < self.importance_threshold],
                remaining
            )
        else:
            others = [m for m in old if getattr(m, 'importance', 0.5) < self.importance_threshold]

        memories.memories = important + others + recent

File: src/AI/middleware/memory_summarization.ts
/**
 * 记忆总结中间件
 * 用于总结智能体的记忆，保持记忆的简洁性和相关性
 */
import { BaseMiddleware, MiddlewareContext, MiddlewareResult } from './base';

export class MemorySummarizationMiddleware extends BaseMiddleware {
  /**
   * 构造函数
   */
  constructor() {
    super("MemorySummarizationMiddleware");
  }

  /**
   * 处理方法
   * 检查是否需要总结记忆
   * @param context 中间件上下文
   * @param agent 智能体实例
   * @returns 中间件结果
   */
  process(context: MiddlewareContext, agent: any): MiddlewareResult {
    // 检查是否需要总结记忆
    // 这里可以实现记忆总结的逻辑，例如：
    // 1. 检测记忆是否过长
    // 2. 对记忆进行摘要
    // 3. 保留重要信息，丢弃不重要的信息
    
    // 目前是一个空实现，返回继续执行
    return {
      should_continue: true,
      metadata: {}
    };
  }
}
